taskchampion/src/storage/test.rs

```diff
@@ -267,6 +272,54 @@ pub(super) fn delete_task_exists(mut storage: impl Storage) -> Result<()> {
     Ok(())
 }
 
+pub(super) fn get_pending_tasks(mut storage: impl Storage) -> Result<()> {
+    let uuid1 = Uuid::new_v4();
+    let uuid2 = Uuid::new_v4();
+
+    {
+        let mut txn = storage.txn()?;
+        txn.add_to_working_set(uuid1)?;
+        txn.add_to_working_set(uuid2)?;
+        txn.commit()?;
+    }
+
+    {
+        let mut txn = storage.txn()?;
+        assert!(txn.create_task(uuid1)?);
+        txn.set_task(
+            uuid1,
+            taskmap_with(vec![("num".to_string(), "1".to_string())]),
+        )?;
+        assert!(txn.create_task(uuid2)?);
+        txn.set_task(
+            uuid2,
+            taskmap_with(vec![("num".to_string(), "2".to_string())]),
+        )?;
+        txn.commit()?;
+    }
+
+    {
+        let mut txn = storage.txn()?;
+        let mut tasks = txn.get_pending_tasks()?;
+
+        let mut exp = vec![
+            (
+                uuid1,
+                taskmap_with(vec![("num".to_string(), "1".to_string())]),
+            ),
+            (
+                uuid2,
+                taskmap_with(vec![("num".to_string(), "2".to_string())]),
+            ),
+        ];
+        exp.sort_by(|a, b| a.0.cmp(&b.0));
+        tasks.sort_by(|a, b| a.0.cmp(&b.0));
+
+        assert_eq!(tasks, exp);
+    }
```

Comment:
I modified this test as follows, and it fails for sqlite. I think this is because of the `LEFT JOIN` - a task which matches no rows in the working set is included in the result set. I think the query should just be `JOIN` (and I think I led you astray in an earlier comment -- sorry!)

```diff
diff --git taskchampion/src/storage/test.rs taskchampion/src/storage/test.rs
index 390bf2f6c..9fe503fd3 100644
--- taskchampion/src/storage/test.rs
+++ taskchampion/src/storage/test.rs
@@ -268,38 +268,41 @@ pub(super) fn delete_task_exists(mut storage: impl Storage) -> Result<()> {
     {
         let mut txn = storage.txn()?;
         assert!(txn.delete_task(uuid)?);
     }
     Ok(())
 }
 
 pub(super) fn get_pending_tasks(mut storage: impl Storage) -> Result<()> {
-    let uuid1 = Uuid::new_v4();
-    let uuid2 = Uuid::new_v4();
+    let uuid1 = Uuid::new_v4(); // in working set
+    let uuid2 = Uuid::new_v4(); // in working set
+    let uuid3 = Uuid::new_v4(); // in working set but does not exist
+    let uuid4 = Uuid::new_v4(); // not in working set but exists
 
     {
         let mut txn = storage.txn()?;
         txn.add_to_working_set(uuid1)?;
         txn.add_to_working_set(uuid2)?;
-        txn.commit()?;
-    }
-
-    {
-        let mut txn = storage.txn()?;
+        txn.add_to_working_set(uuid3)?;
         assert!(txn.create_task(uuid1)?);
         txn.set_task(
             uuid1,
             taskmap_with(vec![("num".to_string(), "1".to_string())]),
         )?;
         assert!(txn.create_task(uuid2)?);
         txn.set_task(
             uuid2,
             taskmap_with(vec![("num".to_string(), "2".to_string())]),
         )?;
+        assert!(txn.create_task(uuid4)?);
+        txn.set_task(
+            uuid4,
+            taskmap_with(vec![("num".to_string(), "4".to_string())]),
+        )?;
         txn.commit()?;
     }
 
     {
         let mut txn = storage.txn()?;
         let mut tasks = txn.get_pending_tasks()?;
 
         let mut exp = vec![
```

Reply:
That was entirely my fault. Your original comment did suggest to use `JOIN`, but I guess I was a bit tired when I wrote it and I didn't make sure I understood the code-base well enough

In any case; thank you for really thinking through the impact of my coding choices, so I can avoid introducing a bug that would frustrate your users :raised_hands: